devtools/working-with-dsproxy/working-with-dsproxy.md

```diff
@@ -0,0 +1,437 @@
+
+}
+```
+
+Let's add the required interfaces to be able to interact with functions on those contracts later in the script.
+
+Add the `TubLike` interface to interact with CDPs on the Tub contract
+
+```text
+interface TubLike {
+    function wipe(bytes32, uint) external;
+    function gov() external view returns (TokenLike);
+    function sai() external view returns (TokenLike);
+    function tab(bytes32) external returns (uint);
+    function rap(bytes32) external returns (uint);
+    function pep() external view returns (PepLike);
+}
+```
+
+Add the `TokenLike` interface to interact with functions on DAI and MKR tokens
+
+```text
+interface TokenLike {
+    function allowance(address, address) external view returns (uint);
+    function balanceOf(address) external view returns (uint);
+    function approve(address, uint) external;
+    function transfer(address, uint) external returns (bool);
+    function transferFrom(address, address, uint) external returns (bool);
+}
+```
+
+Add the `PepLike` interface to read the MKRUSD price from the Pep contract
+
+```text
+interface PepLike {
+    function peek() external returns (bytes32, bool);
+}
+```
+
+Add the `UniswapExchangeLike` interface to be able to retrieve output prices of token swaps and execute them on the Uniswap exchange contracts setup for both DAI and MKR tokens
+
+```text
+interface UniswapExchangeLike {
+    function getEthToTokenOutputPrice(uint256 tokens_bought) external view returns (uint256 eth_sold);
+    function getTokenToEthOutputPrice(uint256 eth_bought) external view returns (uint256 tokens_sold);
+    function tokenToTokenSwapOutput(uint256 tokens_bought, uint256 max_tokens_sold, uint256 max_eth_sold, uint256 deadline, address token_addr) external returns (uint256  tokens_sold);
+}
+```
+
+#### Script outline
+
+Add a new function `wipeWithDai` which takes in the following inputs,
+
+* Address of the Tub contract
+* Address of the Uniswap DAI exchange contract
+* Address of the Uniswap MKR exchange contract
+* Id of the CDP in decimals. Ex: 44
+* Amount of Dai debt to pay back on the CDP
+
+```text
+function wipeWithDai(
+    address _tub,
+    address _DAIExchange,
+    address _MKRExchange,
+    uint cupid,
+    uint wad
+) 
+    public 
+{
+    // logic
+}
+```
+
+#### Checks
+
+Ensure at least some Dai debt is being wiped in the transaction using a require statement
+
+```text
+require(wad > 0);
+```
+
+#### Initialize variables
+
+Initialize contracts using input addresses to interact with their functions later
+
+```text
+TubLike tub = TubLike(_tub);
+UniswapExchangeLike daiEx = UniswapExchangeLike(_DAIExchange);
+UniswapExchangeLike mkrEx = UniswapExchangeLike(_MKRExchange);
+
+TokenLike dai = tub.sai();
+TokenLike mkr = tub.gov();
+PepLike pep =   tub.pep();
+```
+
+Convert the input `cupid` into a left-padded bytes32 hex value format that the Tub contract expects.
+
+```text
+bytes32 cup = bytes32(cupid);
+```
+
+#### Set all allowances
```

Comment:
```suggestion
#### Setup `setAllowance` function
```

Reply:
Since `setAllowance( )` has no dependencies on other functions internal the contract, I think it should be placed before the `wipeWithDai( )` section.